File: examples/c/io-disk-raw/io-disk-raw.c
/* Copyright (c) 2006-2020. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

#include "simgrid/actor.h"
#include "simgrid/disk.h"
#include "simgrid/engine.h"
#include "simgrid/forward.h"
#include "simgrid/host.h"
#include "xbt/dict.h"
#include "xbt/log.h"
#include "xbt/sysdep.h"

#include <stddef.h>
#include <stdlib.h>

XBT_LOG_NEW_DEFAULT_CATEGORY(disk, "Messages specific for this simulation");

static void host(XBT_ATTRIB_UNUSED int argc, XBT_ATTRIB_UNUSED char* argv[])
{
  const char* host_name = sg_host_get_name(sg_host_self());

  /* - Display information on the disks mounted by the current host */
  XBT_INFO("*** Storage info on %s ***", host_name);

  /* - Retrieve all disks from current host */
  unsigned int disk_count;
  sg_disk_t* disk_list;
  sg_host_disks(sg_host_self(), &disk_count, &disk_list);

  for (unsigned int i = 0; i < disk_count; i++)
    XBT_INFO("Disk name: %s (read: %.0f B/s -- write: %.0f B/s ", sg_disk_name(disk_list[i]),
             sg_disk_read_bandwidth(disk_list[i]), sg_disk_write_bandwidth(disk_list[i]));

  /* - Write 400,000 bytes on Disk1 */
  sg_disk_t disk  = disk_list[0];
  sg_size_t write = sg_disk_write(disk, 400000);
  XBT_INFO("Wrote %llu bytes on '%s'", write, sg_disk_name(disk));

  /*  - Now read 200,000 bytes */
  sg_size_t read = sg_disk_read(disk, 200000);
  XBT_INFO("Read %llu bytes on '%s'", read, sg_disk_name(disk));

  /* - Attach some user data to disk1 */
  XBT_INFO("*** Get/set data for storage element: Disk1 ***");

  char* data = (char*)sg_disk_data(disk);

  XBT_INFO("Get storage data: '%s'", data ? data : "No user data");

  sg_disk_data_set(disk, xbt_strdup("Some user data"));
  data = (char*)sg_disk_data(disk);
  XBT_INFO("Set and get data: '%s'", data);
  free(data);
  free(disk_list);
}

int main(int argc, char* argv[])
{
  simgrid_init(&argc, argv);
  simgrid_load_platform(argv[1]);

  simgrid_register_function("host", host);

  size_t host_count = sg_host_count();
  sg_host_t* hosts  = sg_host_list();

  for (long i = 0; i < host_count; i++) {
    XBT_INFO("*** %s properties ****", sg_host_get_name(hosts[i]));
    xbt_dict_t props         = sg_host_get_properties(hosts[i]);
    xbt_dict_cursor_t cursor = NULL;
    char* key;
    void* data;
    xbt_dict_foreach (props, cursor, key, data)
      XBT_INFO("  %s -> %s", key, (char*)data);
    xbt_dict_free(&props);
  }

  free(hosts);

  sg_actor_create("", sg_host_by_name("bob"), host, 0, NULL);

  simgrid_run();

  XBT_INFO("Simulated time %g", simgrid_get_clock());

  return 0;
}

File: examples/java/io_disk_raw/io_disk_raw.java
/* Copyright (c) 2017-2025. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

import org.simgrid.s4u.*;

class HostRunner extends Actor {
  public void run()
  {

    /* -Add an extra disk in a programmatic way */
    Host.current().add_disk("Disk3", /*read bandwidth*/ 9.6e7, /*write bandwidth*/ 6.4e7).seal();

    /* - Display information on the disks mounted by the current host */
    Engine.info("*** Storage info on %s ***", Host.current().get_name());

    /* - Retrieve all disks from current host */
    Disk[] disk_list = Host.current().get_disks();

    /* - For each disk mounted on host, display disk name and mount point */
    for (Disk disk : disk_list)
      Engine.info("Disk name: %s (read: %.0f B/s -- write: %.0f B/s)", disk.get_name(), disk.get_read_bandwidth(),
                  disk.get_write_bandwidth());

    /* - Write 400,000 bytes on Disk1 */
    Disk disk = disk_list[0];
    long write = disk.write(400000);
    Engine.info("Wrote %d bytes on '%s'", write, disk.get_name());

    /*  - Now read 200,000 bytes */
    long read = disk.read(200000);
    Engine.info("Read %d bytes on '%s'", read, disk.get_name());

    /* - Write 800,000 bytes on Disk3 */
    Disk disk3         = disk_list[disk_list.length - 1];
    long write_on_disk3 = disk3.write(800000);
    Engine.info("Wrote %d bytes on '%s'", write_on_disk3, disk3.get_name());

    /* - Attach some user data to disk1 */
    Engine.info("*** Get/set data for storage element: Disk1 ***");

    String data = (String)disk.get_data();

    Engine.info("Get storage data: '%s'", data);

    disk.set_data("Some user data");
    data = (String)disk.get_data();
    Engine.info("Set and get data: '%s'", data);
  }
}

public class io_disk_raw {
  public static void main(String[] args)
  {
    Engine e = new Engine(args);
    e.load_platform(args[0]);

    /* - Display Host properties */
    for (Host h : e.get_all_hosts()) {
      Engine.info("*** %s properties ****", h.get_name());
      for (String key : h.get_properties_names())
        Engine.info("  %s -> %s", key, h.get_property(key));
    }

    e.host_by_name("bob").add_actor("", new HostRunner());

    e.run();
    Engine.info("Simulated ends.");

    // The following call is useless in your code, but our continuous integration uses it to track memleaks
    e.force_garbage_collection();
  }
}
